### PR TITLE
PayPal external service

### PR DESCRIPTION
**Pull request by iker-barriocanal**
# Summary
The system will have two external services as payment methods, being PayPal one of them. The clients of this service will be able to register and pay. The code of this simple service can be divided into two main parts.

## Data
`User` and `Payment` are the data models this server will work with. Since it is a small service, no database interaction has been created, and it is emulated by `DataBase` (that means that after restarting the PayPal server no data will be stored).

### Threads
Various threads compose the server execution. Clients only have one chance to introduce the data on each connection; if the data is wrong, the server will return an error and a new connection must be created.

1. `ServerListener`. Its only purpose is to wait for new clients who want to pay or register in the system. After they choose an option, new threads will be created to process that functionality, so that `ServerListener` can continue waiting.
2. `Registrator`. Created to register users. When the registration has finished, the thread is finished.
3. `Payer`. Created to allow users make payments. When the payment has finished, the thread is finished.

### Main class
`PayPal` is the main class of the server. It initializes a `ServerListener` and keeps track of the threads in ready or running state in the server. When a thread finishes, it removes itself automatically from the pool in this class, but they must be added (threads don't add themselves to the pools). Since there is a single pool for each type of thread, there's also a mutual exclusive lock to make those pools thread-safe.

### CODE REVIEW

**Review by iapellaniz — COMMENTED** [review]
Registrator is not an english word! Maybe "registrar" is a good alternative? (should be easy to rename, if not, I can help)

Also, could you provide an example implementation? At least a walkthrough for when we have to implement it with the actual Server.

Finally, I'm getting an error when running the tests:

````java
java.net.BindException: Address already in use (Bind failed)
	at java.base/java.net.PlainSocketImpl.socketBind(Native Method)
	at java.base/java.net.AbstractPlainSocketImpl.bind(AbstractPlainSocketImpl.java:437)
	at java.base/java.net.ServerSocket.bind(ServerSocket.java:381)
	at java.base/java.net.ServerSocket.<init>(ServerSocket.java:243)
	at java.base/java.net.ServerSocket.<init>(ServerSocket.java:135)
	at es.deusto.spq.payment.PayPal.connections.ServerListener.<init>(ServerListener.java:55)
	at es.deusto.spq.payment.PayPal.connections.ServerListenerTest.initialize(ServerListenerTest.java:25)
	at java.base/jdk.internal.reflect.NativeMethodAccessorImpl.invoke0(Native Method)
	at java.base/jdk.internal.reflect.NativeMethodAccessorImpl.invoke(NativeMethodAccessorImpl.java:62)
	at java.base/jdk.internal.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
	at java.base/java.lang.reflect.Method.invoke(Method.java:564)
	at org.junit.runners.model.FrameworkMethod$1.runReflectiveCall(FrameworkMethod.java:50)
	at org.junit.internal.runners.model.ReflectiveCallable.run(ReflectiveCallable.java:12)
	at org.junit.runners.model.FrameworkMethod.invokeExplosively(FrameworkMethod.java:47)
	at org.junit.internal.runners.statements.RunBefores.evaluate(RunBefores.java:24)
	at org.junit.runners.ParentRunner.run(ParentRunner.java:363)
	at org.apache.maven.surefire.junit4.JUnit4Provider.execute(JUnit4Provider.java:252)
	at org.apache.maven.surefire.junit4.JUnit4Provider.executeTestSet(JUnit4Provider.java:141)
	at org.apache.maven.surefire.junit4.JUnit4Provider.invoke(JUnit4Provider.java:112)
	at java.base/jdk.internal.reflect.NativeMethodAccessorImpl.invoke0(Native Method)
	at java.base/jdk.internal.reflect.NativeMethodAccessorImpl.invoke(NativeMethodAccessorImpl.java:62)
	at java.base/jdk.internal.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
	at java.base/java.lang.reflect.Method.invoke(Method.java:564)
	at org.apache.maven.surefire.util.ReflectionUtils.invokeMethodWithArray(ReflectionUtils.java:189)
	at org.apache.maven.surefire.booter.ProviderFactory$ProviderProxy.invoke(ProviderFactory.java:165)
	at org.apache.maven.surefire.booter.ProviderFactory.invokeProvider(ProviderFactory.java:85)
	at org.apache.maven.surefire.booter.ForkedBooter.runSuitesInProcess(ForkedBooter.java:115)
	at org.apache.maven.surefire.booter.ForkedBooter.main(ForkedBooter.java:75)
````
I know they work on CI, but after restarting my computer, the error seems to persist.

**Comment by iker-barriocanal** [comment]
## Requirements
In order to implement it in the server (or any other place), the requirements are:
1. Have network connection to the PayPal server.
2. Follow the workflow language.

## Workflow
The workflow starts sending a first message of "REGISTER" or "PAY" to the server; then, depending on which option is taken, the flow will be different. However, if the connection is established and then is switched off, the server has been closed.

### Register option
1. Server asks "USERNAME". The client must send it.
2. Server asks "PASSWORD". The client must send it.
3. Optionally, server asks "QUANTITY". The client must send it.
4. If the data has been processed successfully, server returns an "OK" and switches off the connetion. If not, server returns "ERROR" and also finishes the connection.

### Pay option
1. Server asks "USERNAME". The client must send it.
2. Server asks "PASSWORD". The client must send it.
3. Server asks "PRICE". The client must send it.
4. If the data has been processed successfully, server returns an "OK" and switches off the connetion. If not, server returns "ERROR" and also finishes the connection.

File: src/test/java/es/deusto/spq/payment/PayPal/connections/ServerListenerTest.java [hunk 1]
```diff
@@ -0,0 +1,41 @@
...
+	private static PayPal paypal;
+	
+	/**
+	 * Initializes the necessary data to execute the test.
+	 * @throws IOException - launched by {@code java.net}.
+	 */
+	@BeforeClass
+	public static void initialize() throws IOException {
+		paypal = new PayPal();
+		serverListener = new ServerListener(45000);
```
Comment (iapellaniz) [inline src/test/java/es/deusto/spq/payment/PayPal/connections/ServerListenerTest.java]: Tests seem to be conflicting with the use of the same port. What about changing it for this one to, say, 45001?
That way it doesn't fail on my end.

**Review by iapellaniz — APPROVED** [review]
Good!

**Pull request merged**